File: 5회차-linearDS/5회차 필수문제 코드/CPP/14713.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n; cin >> n;
    cin.ignore();

    stack<string> stk[105];
    for (int i = 0; i <= n; ++i) {
        string s;
        getline(cin, s);

        int pos = 0;
        for (int j = 0; j < s.length(); ++j) {
            if (s[j] == ' ') {
                stk[i].push(s.substr(pos, j - pos));
                pos = j + 1;
            }
        }
        stk[i].push(s.substr(pos));
    }

    bool flag = true;

    while (!stk[n].empty()) {
        string now = stk[n].top();
        stk[n].pop();

        bool find = false;
        for (int i = 0; i < n; ++i) {
            if (stk[i].empty()) continue;
            if (stk[i].top() != now) 
                continue;

            stk[i].pop();
            find = true;
        }

        if (find == false) {
            flag = false;
            break;
        }
    }

    if (flag == true)
        cout << "Possible" << '\n';
    else
        cout << "Impossible" << '\n';
}

File: 5회차-linearDS/5회차 필수문제 코드/PYTHON/14713.py
n = int(input())
stk = [[] for _ in range(105)]

for i in range(n + 1):
    s = input()
    pos = 0
    for j in range(len(s)):
        if s[j] == ' ':
            stk[i].append(s[pos : j])
            pos = j + 1
    stk[i].append(s[pos:])

flag = True

while len(stk[n]) != 0:
    now = stk[n].pop()
    find = False
    for i in range(n):
        if len(stk[i]) == 0 or stk[i][-1] != now:
            continue

        stk[i].pop()
        find = True
    
    if find == False:
        flag = False
        break

if flag == True:
    print("Possible")
else:
    print("Impossible")
